packages/ra-ui-materialui/src/list/SimpleList.tsx

```diff
@@ -271,35 +280,52 @@ const LinkOrNot = (props: LinkOrNotProps) => {
         id,
         children,
         record,
+        resource,
     } = props;
     const classes = useLinkOrNotStyles({ classes: classesOverride });
-    const link =
-        typeof linkType === 'function' ? linkType(record, id) : linkType;
+    const [effect, setEffect] = useState<string>(() =>
+        linkType === 'edit' || linkType === true
+            ? linkToRecord(basePath || `/${resource}`, id)
+            : linkToRecord(basePath || `/${resource}`, id, 'show')
+    );
 
-    return link === 'edit' || link === true ? (
-        <Link to={linkToRecord(basePath, id)} className={classes.link}>
-            {children}
-        </Link>
-    ) : link === 'show' ? (
-        <Link
-            to={`${linkToRecord(basePath, id)}/show`}
-            className={classes.link}
-        >
+    useEffect(() => {
+        if (typeof linkType !== 'function') {
+            return;
+        }
+        const getEffect = async () => {
+            if (typeof linkType === 'function') {
+                setEffect(
+                    await linkType(record, id, basePath || `/${resource}`)
```

Comment:
why assume that this is async? It's not the case in Datagrid and it complexifies everything.

Reply:
https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/list/datagrid/DatagridRow.tsx#L113

Reply:
Then why don't you do it in a click handler as in the datagrid row?

Reply:
Wanted to keep the link